Flaky test: TestDisableUserData_QueryTimesOut

## Expected Behavior
Test should always pass

## Actual Behavior
```
--- FAIL: TestVersioningIntegrationSuite (15.42s)
    --- FAIL: TestVersioningIntegrationSuite/TestDisableUserData_QueryTimesOut (0.04s)
        versioning_test.go:1631: 
                Error Trace:    /Users/ycyang/temporalio/temporal/tests/versioning_test.go:1631
                Error:          Should be in error chain:
                                expected: %!q(**serviceerror.DeadlineExceeded=0x14000c9e780)
                                in chain: 
                Test:           TestVersioningIntegrationSuite/TestDisableUserData_QueryTimesOut
```

## Steps to Reproduce the Problem

  1. Run ^ test using master branch (I was using commit `7e2de6902fa98cc98884b4f4003b5a0e8a91936e`). It fails in like 1-2 out of 10 runs.
  1.
  1.

## Specifications

  - Version:
  - Platform:

--- a/tests/versioning_test.go
+++ b/tests/versioning_test.go
@@ -1591,19 +1591,28 @@ func (s *versioningIntegSuite) TestDisableUserData_WorkflowGetsStuck() {
 	s.Require().Equal(int32(1), runs.Load())
 }
 
-func (s *versioningIntegSuite) TestDisableUserData_QueryTimesOut() {
+func (s *versioningIntegSuite) TestDisableUserData_QueryFails() {
+	// force one partition so that we can unload the task queue
+	dc := s.testCluster.host.dcClient
+	dc.OverrideValue(dynamicconfig.MatchingNumTaskqueueReadPartitions, 1)
+	dc.OverrideValue(dynamicconfig.MatchingNumTaskqueueWritePartitions, 1)
+	defer dc.RemoveOverride(dynamicconfig.MatchingNumTaskqueueReadPartitions)
+	defer dc.RemoveOverride(dynamicconfig.MatchingNumTaskqueueWritePartitions)
+
 	tq := s.T().Name()
 	v1 := s.prefixed("v1")
 	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 	defer cancel()
+
 	s.addNewDefaultBuildId(ctx, tq, v1)
 
 	var runs atomic.Int32
 	wf := func(ctx workflow.Context) error {
-		return workflow.SetQueryHandler(ctx, "query", func() (string, error) {
+		workflow.SetQueryHandler(ctx, "query", func() (string, error) {
 			runs.Add(1)
 			return "response", nil
 		})
+		return nil
 	}
 	wrk := worker.New(s.sdkClient, tq, worker.Options{
 		BuildID:                          v1,
@@ -1620,16 +1629,18 @@ func (s *versioningIntegSuite) TestDisableUserData_QueryTimesOut() {
 	}, "wf")
 	s.Require().NoError(err)
 
-	dc := s.testCluster.host.dcClient
-	defer dc.RemoveOverride(dynamicconfig.MatchingLoadUserData)
+	// wait for it to complete
+	s.NoError(run.Get(ctx, nil))
+
 	dc.OverrideValue(dynamicconfig.MatchingLoadUserData, false)
+	defer dc.RemoveOverride(dynamicconfig.MatchingLoadUserData)
 
 	s.unloadTaskQueue(ctx, tq)
 
 	_, err = s.sdkClient.QueryWorkflow(ctx, run.GetID(), run.GetRunID(), "query")
-	var deadlineExceededError *serviceerror.DeadlineExceeded
-	s.Require().ErrorAs(err, &deadlineExceededError)
-	s.Require().Equal(int32(0), runs.Load())
+	var failedPrecond *serviceerror.FailedPrecondition
+	s.ErrorAs(err, &failedPrecond, err)
+	s.Equal(int32(0), runs.Load())
 }
 
 func (s *versioningIntegSuite) TestDescribeTaskQueue() {

--- a/tests/versioning_test.go
+++ b/tests/versioning_test.go
@@ -1539,15 +1539,21 @@ func (s *versioningIntegSuite) TestDisableUserData_UnversionedWorkflowRuns() {
 }
 
 func (s *versioningIntegSuite) TestDisableUserData_WorkflowGetsStuck() {
+	// force one partition so that we can unload the task queue
+	dc := s.testCluster.host.dcClient
+	dc.OverrideValue(dynamicconfig.MatchingNumTaskqueueReadPartitions, 1)
+	dc.OverrideValue(dynamicconfig.MatchingNumTaskqueueWritePartitions, 1)
+	defer dc.RemoveOverride(dynamicconfig.MatchingNumTaskqueueReadPartitions)
+	defer dc.RemoveOverride(dynamicconfig.MatchingNumTaskqueueWritePartitions)
+
 	tq := s.T().Name()
 	v1 := s.prefixed("v1")
 	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 	defer cancel()
 	s.addNewDefaultBuildId(ctx, tq, v1)
 
-	dc := s.testCluster.host.dcClient
-	defer dc.RemoveOverride(dynamicconfig.MatchingLoadUserData)
 	dc.OverrideValue(dynamicconfig.MatchingLoadUserData, false)
+	defer dc.RemoveOverride(dynamicconfig.MatchingLoadUserData)
 
 	s.unloadTaskQueue(ctx, tq)
 

--- a/tests/versioning_test.go
+++ b/tests/versioning_test.go
@@ -399,6 +399,7 @@ func (s *versioningIntegSuite) dispatchNewWorkflowStartWorkerFirst() {
 }
 
 func (s *versioningIntegSuite) TestDisableUserData_DefaultTasksBecomeUnversioned() {
+	// force one partition so that we can unload the task queue
 	dc := s.testCluster.host.dcClient
 	dc.OverrideValue(dynamicconfig.MatchingNumTaskqueueReadPartitions, 1)
 	defer dc.RemoveOverride(dynamicconfig.MatchingNumTaskqueueReadPartitions)
@@ -1481,13 +1482,11 @@ func (s *versioningIntegSuite) TestDisableUserData() {
 	// First insert some data (we'll try to read it below)
 	s.addNewDefaultBuildId(ctx, tq, v1)
 
-	// unload so that we reload and pick up LoadUserData dynamic config
-	s.unloadTaskQueue(ctx, tq)
-
 	dc := s.testCluster.host.dcClient
 	defer dc.RemoveOverride(dynamicconfig.MatchingLoadUserData)
 	dc.OverrideValue(dynamicconfig.MatchingLoadUserData, false)
 
+	// unload so that we reload and pick up LoadUserData dynamic config
 	s.unloadTaskQueue(ctx, tq)
 
 	// Verify update fails

